### PR TITLE
TPF objects can now be added, subtracted, multiplied and divided by numpy arrays

### PR DESCRIPTION
**Pull request by christinahedges**
It's useful to be able to do simple maths with TPF objects in order to e.g. remove background. I've added the capability to do that.

There are special rules depending on the shape of the object you try to add/subtract/divide/multiply. 

- If it's a point, it's add/subtract/divide/multiply-ed from all values.
- If it's a vector with **length time** , it's add/subtract/divide/multiply-ed from each frame.
- If it's a vector **NOT of length time**  it raises a value error
- If it's a 2D array with **shape npixels x mpixels** , it's add/subtract/divide/multiply-ed from each pixel.
- If it's a 2D array with **shape NOT npixels x mpixels** , it raises a value error
- If it's a 3D array with **shape time x npixels x mpixels** , it's add/subtract/divide/multiply-ed from each value.
- If it's a 3D array with **shape NOT time x npixels x mpixels** , it raises a value error

image

There are lovely accompanying tests.

### CODE REVIEW

File: lightkurve/tests/test_targetpixelfile.py [hunk 1]
```diff
@@ -60,6 +60,41 @@ def test_tpf_shapes():
...
+        warnings.simplefilter("ignore", LightkurveWarning)
+        tpfs = [KeplerTargetPixelFile(filename_tpf_all_zeros),
+                TessTargetPixelFile(filename_tpf_all_zeros)]
+    # These should work
+    for tpf in tpfs:
+        for other in [1, [1], np.arange(len(tpf.time)), np.ones(tpf.flux.shape[1:]), np.ones(tpf.shape)]:
+            tpf + other
+            tpf - other
+            tpf * other
+            tpf / other
```
Comment (barentsen) [inline lightkurve/tests/test_targetpixelfile.py]: Codacy Issue found: [Statement seems to have no effect](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4836899)

File: lightkurve/tests/test_targetpixelfile.py [hunk 2]
```diff
@@ -60,6 +60,41 @@ def test_tpf_shapes():
...
+    with warnings.catch_warnings():
+        # Ignore the "TELESCOP is not equal to TESS" warning
+        warnings.simplefilter("ignore", LightkurveWarning)
+        tpfs = [KeplerTargetPixelFile(filename_tpf_all_zeros),
+                TessTargetPixelFile(filename_tpf_all_zeros)]
+    # These should work
+    for tpf in tpfs:
+        for other in [1, [1], np.arange(len(tpf.time)), np.ones(tpf.flux.shape[1:]), np.ones(tpf.shape)]:
+            tpf + other
+            tpf - other
```
Comment (barentsen) [inline lightkurve/tests/test_targetpixelfile.py]: Codacy Issue found: [Statement seems to have no effect](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4836899)

File: lightkurve/tests/test_targetpixelfile.py [hunk 3]
```diff
@@ -60,6 +60,41 @@ def test_tpf_shapes():
...
+            tpf * other
```
Comment (barentsen) [inline lightkurve/tests/test_targetpixelfile.py]: Codacy Issue found: [Statement seems to have no effect](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4836899)

File: lightkurve/tests/test_targetpixelfile.py [hunk 4]
```diff
@@ -60,6 +60,41 @@ def test_tpf_shapes():
...
+            tpf += other
+            tpf -= other
+            tpf *= other
+            tpf /= other
+
+    # These should fail with a value error because their shape is wrong.
+    for tpf in tpfs:
+        for other in [[1, 2], np.arange(len(tpf.time) - 1), np.ones([100, 1]), np.ones([1, 2, 3])]:
+            with pytest.raises(ValueError):
+                tpf + other
```
Comment (barentsen) [inline lightkurve/tests/test_targetpixelfile.py]: Codacy Issue found: [Statement seems to have no effect](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4836899)

File: lightkurve/tests/test_targetpixelfile.py (hunk 1 above)
Comment (barentsen) [inline lightkurve/tests/test_targetpixelfile.py]: Codacy Issue found: [Statement seems to have no effect](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4836899)

File: lightkurve/tests/test_targetpixelfile.py (hunk 2 above)
Comment (barentsen) [inline lightkurve/tests/test_targetpixelfile.py]: Codacy Issue found: [Statement seems to have no effect](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4836899)

File: lightkurve/tests/test_targetpixelfile.py (hunk 3 above)
Comment (barentsen) [inline lightkurve/tests/test_targetpixelfile.py]: Codacy Issue found: [Statement seems to have no effect](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4836899)

File: lightkurve/tests/test_targetpixelfile.py (hunk 4 above)
Comment (barentsen) [inline lightkurve/tests/test_targetpixelfile.py]: Codacy Issue found: [Statement seems to have no effect](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4836899)

**Comment by christinahedges** [comment]
@barentsen it looks like the remote data test is failing because the CI isn't allowed to delete the temporary file...

File: lightkurve/targetpixelfile.py [hunk 5]
```diff
@@ -80,6 +82,107 @@ def __getitem__(self, key):
             copy[1].data = copy[1].data[selected_idx]
         return self.__class__(copy, quality_bitmask=self.quality_bitmask, targetid=self.targetid)
 
+    def __len__(self):
+        return len(self.time)
+
+
+    def _verify_other_(self, other):
```
Comment (barentsen) [inline lightkurve/targetpixelfile.py]: Even though it's a helper function, let's have a one-line docstring here to explain what this function is for.

**Comment by barentsen** [comment]
Do we need the custom math logic?  If we used the standard numpy broadcasting, we'd already support a single scalar, a 2D array, and a 3D array in exactly the same way.

The only thing that would be different is math with a vector of length time, which would behave like this:
```python
>>> import numpy as np
>>> np.ones((3,3,3)) - np.array([1,2,3])
array([[[ 0., -1., -2.],
        [ 0., -1., -2.],
        [ 0., -1., -2.]],

       [[ 0., -1., -2.],
        [ 0., -1., -2.],
        [ 0., -1., -2.]],

       [[ 0., -1., -2.],
        [ 0., -1., -2.],
        [ 0., -1., -2.]]])
```

Is there a strong enough argument to override the latter?

**Comment by barentsen** [comment]
> @barentsen it looks like the remote data test is failing because the CI isn't allowed to delete the temporary file...

At least one of the tests seemed to have failed because of a simple MAST download issue.
```
>       for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
E       socket.gaierror: [Errno 8] nodename nor servname provided, or not known. requested URL: http://archive.stsci.edu/missions/kepler/fpc/prf/extracted/kplr13.2_2011265_prf.fits
```

Re-started all the tests just now.

File: lightkurve/tests/test_targetpixelfile.py [hunk 6]
```diff
@@ -60,6 +60,41 @@ def test_tpf_shapes():
...
+        warnings.simplefilter("ignore", LightkurveWarning)
+        tpfs = [KeplerTargetPixelFile(filename_tpf_all_zeros),
+                TessTargetPixelFile(filename_tpf_all_zeros)]
+    # These should work
+    for tpf in tpfs:
+        for other in [1, np.ones(tpf.flux.shape[1:]), np.ones(tpf.shape)]:
+            tpf + other
+            tpf - other
+            tpf * other
+            tpf / other
```
Comment (barentsen) [inline lightkurve/tests/test_targetpixelfile.py]: Codacy Issue found: [Statement seems to have no effect](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4836899)

File: lightkurve/tests/test_targetpixelfile.py [hunk 7]
```diff
@@ -60,6 +60,41 @@ def test_tpf_shapes():
...
+    with warnings.catch_warnings():
+        # Ignore the "TELESCOP is not equal to TESS" warning
+        warnings.simplefilter("ignore", LightkurveWarning)
+        tpfs = [KeplerTargetPixelFile(filename_tpf_all_zeros),
+                TessTargetPixelFile(filename_tpf_all_zeros)]
+    # These should work
+    for tpf in tpfs:
+        for other in [1, np.ones(tpf.flux.shape[1:]), np.ones(tpf.shape)]:
+            tpf + other
+            tpf - other
```
Comment (barentsen) [inline lightkurve/tests/test_targetpixelfile.py]: Codacy Issue found: [Statement seems to have no effect](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4836899)

File: lightkurve/tests/test_targetpixelfile.py [hunk 8]
```diff
@@ -60,6 +60,41 @@ def test_tpf_shapes():
...
+            tpf * other
```
Comment (barentsen) [inline lightkurve/tests/test_targetpixelfile.py]: Codacy Issue found: [Statement seems to have no effect](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4836899)

File: lightkurve/tests/test_targetpixelfile.py [hunk 9]
```diff
@@ -60,6 +60,41 @@ def test_tpf_shapes():
...
+            tpf += other
+            tpf -= other
+            tpf *= other
+            tpf /= other
+
+    # These should fail with a value error because their shape is wrong.
+    for tpf in tpfs:
+        for other in [np.asarray([1, 2]), np.arange(len(tpf.time) - 1), np.ones([100, 1]), np.ones([1, 2, 3])]:
+            with pytest.raises(ValueError):
+                tpf + other
```
Comment (barentsen) [inline lightkurve/tests/test_targetpixelfile.py]: Codacy Issue found: [Statement seems to have no effect](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4836899)

**Comment by barentsen** [comment]
Mmmm, I'm going to restart the unit tests to make sure this isn't a transient problem, but this doesn't appear to work on Windows yet because of the following error:

```
    def __add__(self, other):
        hdu = deepcopy(self.hdu)
        tmp = tempfile.NamedTemporaryFile(delete=False)
        hdu[1].data['FLUX'][self.quality_mask] += other
        fits.HDUList(hdus=list(hdu)).writeto(tmp.name, overwrite=True)
        tpf = type(self)(tmp.name, quality_bitmask=self.quality_bitmask)
        tpf.path = None
        tmp.close()
>       os.remove(tmp.name)
E       PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\tmp5oykpf8g'
```

**Comment by barentsen** [comment]
I rebased on master and force-pushed, because some of the failing unit tests have been fixed on master.  Let's see how we do now!

File: lightkurve/tests/test_targetpixelfile.py (hunk 6 above)
Comment (barentsen) [inline lightkurve/tests/test_targetpixelfile.py]: Codacy Issue found: [Statement seems to have no effect](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4836899)

File: lightkurve/tests/test_targetpixelfile.py (hunk 7 above)
Comment (barentsen) [inline lightkurve/tests/test_targetpixelfile.py]: Codacy Issue found: [Statement seems to have no effect](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4836899)

File: lightkurve/tests/test_targetpixelfile.py (hunk 8 above)
Comment (barentsen) [inline lightkurve/tests/test_targetpixelfile.py]: Codacy Issue found: [Statement seems to have no effect](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4836899)

File: lightkurve/tests/test_targetpixelfile.py (hunk 9 above)
Comment (barentsen) [inline lightkurve/tests/test_targetpixelfile.py]: Codacy Issue found: [Statement seems to have no effect](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4836899)

File: lightkurve/tests/test_targetpixelfile.py (hunk 6 above)
Comment (barentsen) [inline lightkurve/tests/test_targetpixelfile.py]: Codacy Issue found: [Statement seems to have no effect](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4836899)

File: lightkurve/tests/test_targetpixelfile.py (hunk 7 above)
Comment (barentsen) [inline lightkurve/tests/test_targetpixelfile.py]: Codacy Issue found: [Statement seems to have no effect](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4836899)

File: lightkurve/tests/test_targetpixelfile.py (hunk 8 above)
Comment (barentsen) [inline lightkurve/tests/test_targetpixelfile.py]: Codacy Issue found: [Statement seems to have no effect](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4836899)

File: lightkurve/tests/test_targetpixelfile.py (hunk 9 above)
Comment (barentsen) [inline lightkurve/tests/test_targetpixelfile.py]: Codacy Issue found: [Statement seems to have no effect](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4836899)

File: lightkurve/tests/test_targetpixelfile.py (hunk 6 above)
Comment (barentsen) [inline lightkurve/tests/test_targetpixelfile.py]: Codacy Issue found: [Statement seems to have no effect](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4836899)

File: lightkurve/tests/test_targetpixelfile.py (hunk 7 above)
Comment (barentsen) [inline lightkurve/tests/test_targetpixelfile.py]: Codacy Issue found: [Statement seems to have no effect](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4836899)

File: lightkurve/tests/test_targetpixelfile.py (hunk 8 above)
Comment (barentsen) [inline lightkurve/tests/test_targetpixelfile.py]: Codacy Issue found: [Statement seems to have no effect](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4836899)

File: lightkurve/tests/test_targetpixelfile.py (hunk 9 above)
Comment (barentsen) [inline lightkurve/tests/test_targetpixelfile.py]: Codacy Issue found: [Statement seems to have no effect](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4836899)

File: lightkurve/tests/test_targetpixelfile.py (hunk 6 above)
Comment (barentsen) [inline lightkurve/tests/test_targetpixelfile.py]: Codacy Issue found: [Statement seems to have no effect](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4836899)

File: lightkurve/tests/test_targetpixelfile.py (hunk 7 above)
Comment (barentsen) [inline lightkurve/tests/test_targetpixelfile.py]: Codacy Issue found: [Statement seems to have no effect](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4836899)

File: lightkurve/tests/test_targetpixelfile.py (hunk 8 above)
Comment (barentsen) [inline lightkurve/tests/test_targetpixelfile.py]: Codacy Issue found: [Statement seems to have no effect](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4836899)

File: lightkurve/tests/test_targetpixelfile.py (hunk 9 above)
Comment (barentsen) [inline lightkurve/tests/test_targetpixelfile.py]: Codacy Issue found: [Statement seems to have no effect](https://app.codacy.com/manual/barentsen/lightkurve/pullRequest?prid=4836899)

**Comment by barentsen** [comment]
This feature isn't working on Python 2 for an error I don't understand:
```
E       TypeError: Required argument 'fileno' (pos 1) not found
```

I'm really tempted to drop Python 2 now. It's 2020!

**Comment by barentsen** [comment]
Because this doesn't work on Python 2, I've assigned this to the Lightkurve v2.0 milestone.  We can merge when we are preparing that release!

**Comment by barentsen** [comment]
We dropped support for Python 2 in #733 so we can now merge this PR!

I rebased on master.  All tests pass. :+1:  I intend to merge this shortly!

**Pull request merged**